model/src/main/java/com/epam/esm/dao/impl/TagDaoImpl.java

```diff
@@ -1,91 +1,68 @@
 package com.epam.esm.dao.impl;
 
-import com.epam.esm.dao.TagDAO;
-import com.epam.esm.dao.mapper.TagMapper;
+import com.epam.esm.dao.TagDao;
 import com.epam.esm.entity.Tag;
-import com.epam.esm.exception.DaoException;
-import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.dao.DataAccessException;
-import org.springframework.jdbc.core.JdbcTemplate;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
 
 import java.util.List;
-
-import static com.epam.esm.exception.ExceptionDaoMessageCodes.*;
+import java.util.Optional;
 
 @Repository
-public class TagDaoImpl implements TagDAO {
+@Transactional
+public class TagDaoImpl implements TagDao {
 
-    private final String QUERY_CREATE_TAG = "insert into tags (tag_name) values(?)";
-    private final String QUERY_GET_TAG = "select * from tags where id = ?;";
-    private final String QUERY_GET_TAG_ALL = "select * from tags;";
-    private final String QUERY_GET_BY_NAME = "select * from tags where tag_name = ?";
-    private final String QUERY_DELETE_BY_ID = "delete from tags where id = ?";
+    private static final String QUERY_GET_TAG_ALL = "select t from Tag as t";
+    private static final String QUERY_GET_BY_NAME = "select t from Tag as t where t.name = :name";
+    private static final String QUERY_SELECT_MOST_POPULAR_TAG = "select t from Order o " +
```

Comment:
Seems to me, this query is wrong. In the task, it is written like that
**Get the most widely used tag of a user with the highest cost of all orders**
It means that API should return the tag which is used widely by one exact user and has the highest cost of all orders with this tag.
Imagine there are these tables:
Tag table
id - 1, name - tech
id - 2, name - book
id -3, name - gym

Gift certificate table
id - 1, price - 1000, tagId - 1
id - 2, price - 2000, tagId - 1
id - 3, price - 3000, tagId - 1
id - 4, price - 3000, tagId - 2
id - 5, price - 2000, tagId - 2
id - 6, price - 1000, tagId - 2
id - 7, price - 4000, tagId - 3

orders
id - 1, cost - 1000, certificateId - 1, userId - 1
id - 2, cost - 2000, certificateId - 2, userId - 1
id - 3, cost - 3000, certificateId - 3, userId - 1
id - 4, cost - 3000, certificateId - 4, userId - 1
id - 5, cost - 2000, certificateId - 5, userId - 1
id - 6, cost - 1000, certificateId - 6, userId - 1
id - 7, cost - 4000, certificateId - 7, userId - 1

User with id 1, bought 7 certificates. The most widely used tags are tech and book, three times. Their prices are the same, 6000. So both of them should be returned.